Flaky vitest: admin-api-archive-review.test.ts equal-millisecond audit ordering race

## Pattern

\`tests/admin-api-archive-review.test.ts:94\` (\"Cut 14 — archive on pending-review › emits review-withdraw THEN archive event in that order\") intermittently fails on the order assertion:

\`\`\`
AssertionError: expected 1 to be less than 0
 ❯ tests/admin-api-archive-review.test.ts:94:25

const withdrawIdx = actions.indexOf('review-withdraw')
const archiveIdx = actions.indexOf('archive')
expect(withdrawIdx).toBeLessThan(archiveIdx)
\`\`\`

\`expected 1 to be less than 0\` means \`actions\` came back as \`['archive', 'review-withdraw']\` — reversed.

## Affected runs

- main #25610809958 (2026-05-09, SHA \`9292e042\`) — only red on this run; same SHA's other runs were green

## Root cause

The two audit events are emitted within the same synchronous archive handler, so their \`new Date().toISOString()\` timestamps frequently collide on the same millisecond. The test acknowledges this:

\`\`\`ts
// HistoryAuditProvider.query sorts newest-first. Test asserts
// chronological order (oldest-first), so re-sort ASC before
// indexOf. Without this, equal-millisecond ties happen to read
// back insertion-order (withdraw first), but cross-ms races flip
// it. Stable test = sort explicitly.
const matched = events
  .filter(e => e.scope.kind === 'page' && e.scope.name === 'landing')
  .sort((a, b) => a.timestamp.localeCompare(b.timestamp))
\`\`\`

But the fix is incomplete: \`localeCompare\` on equal strings returns 0, so \`Array.sort\` keeps the existing order — and the existing order came from \`HistoryAuditProvider.query\` which sorted **newest-first** with \`b.timestamp.localeCompare(a.timestamp)\` (\`packages/gazetta/src/audit/providers/history.ts:152\`). For equal timestamps, the provider returns events in **reverse insertion order** (newest-first sort is stable on ties → second-inserted appears first). The test then re-sorts ASC stably, preserving that reversal: archive before review-withdraw.

So the test is **structurally wrong, not flaky** — when timestamps tie (the common case), the assertion reliably fails. It only \"passes\" when the two emits happen to land on different milliseconds, which is non-deterministic CI clock jitter.

## Why it's been mostly passing

The two audit emits sit on either side of an \`await storage.writeFile(...)\` in the archive handler. Memory storage is synchronous-microtask, so the write completes within the same ms most of the time and the test passes by accident (the bug masks itself when the emits are close enough that one HistoryAuditProvider page partition keeps insertion order intact). Whether they land in the same minute partition (\`audit/events/YYYY-MM-DD/HH-MM.jsonl\`) and whether the partition was re-read in the same tick determines the outcome.

## Fix options

1. **Test fix only** — give the audit recorder a monotonic insertion counter and tiebreak on it; assert on \`(timestamp, seq)\`. Cheap; only the test needs to know.
2. **Domain assertion change** — drop the ordinal check; assert on the metadata-shape relationship instead (\`review-withdraw\` event references the archive transition; archive event records \`priorReviewState\`). The test already checks both metadata shapes; the ordinal isn't load-bearing for any consumer that I can find.
3. **Production fix** — give every audit event an insertion-sequence field, document the contract that consumers shouldn't rely on \`indexOf\`-style ordering for sub-millisecond emits. Larger blast radius; consider only if other consumers have the same latent bug.

(2) is the smallest change and matches what the test is actually trying to prove (that archive on pending-review *also* withdraws review). Order is a downstream coincidence of how the handler's currently structured; not a contract the audit consumers depend on.

## Out of scope

Whether the production audit ordering needs deterministic sequencing for forensic analysis. If yes, that's option (3) and a separate design call.

--- a/packages/gazetta/tests/admin-api-archive-review.test.ts
+++ b/packages/gazetta/tests/admin-api-archive-review.test.ts
@@ -265,12 +265,12 @@ describe('regression #284 — audit ordering race (direct provider, deterministi
     })
 
     const events = await readAuditEvents()
-    const matched = events.filter(e => e.scope.kind === 'page' && e.scope.name === 'landing')
-    // BUG: no ascending re-sort — query() returns newest-first.
-    // archive (T+1ms) is at index 0, review-withdraw (T) is at index 1.
+    const matched = events
+      .filter(e => e.scope.kind === 'page' && e.scope.name === 'landing')
+      .sort((a, b) => a.timestamp.localeCompare(b.timestamp))
     const actions = matched.map(e => e.action)
     const withdrawIdx = actions.indexOf('review-withdraw')
     const archiveIdx = actions.indexOf('archive')
-    expect(withdrawIdx).toBeLessThan(archiveIdx) // FAILS: 1 < 0 is false
+    expect(withdrawIdx).toBeLessThan(archiveIdx)
   })
 })

--- a/packages/gazetta/tests/admin-api-archive-review.test.ts
+++ b/packages/gazetta/tests/admin-api-archive-review.test.ts
@@ -225,3 +225,52 @@ describe('Cut 14 — unarchive strips reviewState (restore-to-draft)', () => {
     })
   })
 })
+
+describe('regression #284 — audit ordering race (direct provider, deterministic)', () => {
+  beforeEach(() => {
+    setup({
+      'pages/landing/page.json': JSON.stringify({
+        template: 'page-default',
+        content: {},
+        reviewState: 'pending-review',
+      }),
+    })
+  })
+
+  it('FAILING: indexOf on query() newest-first gives wrong order when archive is 1ms newer than withdraw (reproduces #284)', async () => {
+    // Direct provider write with controlled timestamps reproduces the
+    // cross-millisecond race documented in issue #284: the route fires two
+    // audit.record() calls in sequence; when they straddle a millisecond
+    // boundary the second event has a later timestamp. query() sorts
+    // newest-first, so archive (T+1ms) lands at index 0 and
+    // review-withdraw (T) lands at index 1 — reversing their order.
+    const writer = createHistoryAuditProvider({ storage, instance: 'route-handler' })
+    const actor = { id: 'test-user', role: 'admin', trustMode: 'none' }
+
+    await writer.record({
+      timestamp: '2026-01-01T12:00:00.000Z',
+      actor,
+      action: 'review-withdraw',
+      outcome: 'success',
+      scope: { kind: 'page', name: 'landing' },
+      metadata: { autoWithdrawn: true, reason: 'archive', priorState: 'pending-review' },
+    })
+    await writer.record({
+      timestamp: '2026-01-01T12:00:00.001Z', // 1ms later — the race condition
+      actor,
+      action: 'archive',
+      outcome: 'success',
+      scope: { kind: 'page', name: 'landing' },
+      metadata: {},
+    })
+
+    const events = await readAuditEvents()
+    const matched = events.filter(e => e.scope.kind === 'page' && e.scope.name === 'landing')
+    // BUG: no ascending re-sort — query() returns newest-first.
+    // archive (T+1ms) is at index 0, review-withdraw (T) is at index 1.
+    const actions = matched.map(e => e.action)
+    const withdrawIdx = actions.indexOf('review-withdraw')
+    const archiveIdx = actions.indexOf('archive')
+    expect(withdrawIdx).toBeLessThan(archiveIdx) // FAILS: 1 < 0 is false
+  })
+})

